### PR TITLE
Introduce EMPTY_SPACE terrain flag

### LINKED ISSUE #72930: Provide "empty space" terrain with a property indicating its emptiness.
### Is your feature request related to a problem? Please describe.

I've currently been in contact with code dealing with the question of whether roofs should be generated over terrain that's supposed to have roofs. The current code seems to check for t_open_air, but that's a bit primitive, looking forwards.

We already have various versions of deep water, and I believe there are plans/ideas for underwater structures, plus I can imagine various settings adding additional kinds of "empty space", such as vacuum and non air atmospheres.

Thus, I think it would make sense to provide a means by which we can identify if we have any of these non blocking "empty" tiles and use that property to determine whether we should/could place terrain there, possibly secondary terrain (such as a roof), but also as a check for general emptiness for other purposes.

### Solution you would like.

I'm not too fussed about the form such an indicator would take. I guess a flag would do it, for instance, but it could also be an additional element defaulting to not empty space.

### Describe alternatives you have considered.

Continue to check for t_open_air and leave it for later to expand the logic to handle other cases (whether by checking for a range or tiles or introduce this kind of logic then).

### Additional context

Trees, roofs, holes. Lots of stuff that doesn't work properly because there's either no code for it or because the add_roofs magic relied on isn't magic enough to patch things retroactively in all cases (and patches up things that should not be patched up, in the case of some holes).
There's a need for improvements in these areas, and it makes sense to provide support for doing it properly, rather than leaving it for someone to do it later when things have been fixed for the open air case specifically.

### PR DESCRIPTION
**Pull request by PatrikLundell**
#### Summary
None

#### Purpose of change

Fix #72930, i.e. introduce an indication that a terrain tile is empty of anything solid, and thus cannot act as a roof.

#### Describe the solution
Adding a new JSON flag for terrain to indicate it's empty of anything solid.
This flag is then checked for rather than explicitly checking for t_open_air in the context of player/companion construction.

#### Describe alternatives you've considered
Scope creep by trying to address potentially erroneous uses of direct check for or assignment of t_open_air.
- map.cpp, map::collapse_at: Sets t_open_air, but might need to check if the "rooved" variants should be placed instead.
- map::get_roof: Can return t_open_air, but might need to check if it should return a "rooved" variant instead.
- map::bash_ter_furn: ditto.
- map_field.cpp, field_processor_fd_fire: ditto.
- vehicle_autodrive.cpp, autodrive_controller::check_drivable: Demanding open air for a helo is reasonable, as you probably shouldn't be able to fly under a roof. Auto drive for submarines should be separate anyway.
However, I suspect the code would consider "rooved" open air to be valid for both land and air vehicles.

#### Testing
Loaded a save with an active task to make a hole out of a roof. Verified the terrain beneath is still transformed into a roof less terrain.

#### Additional context
It can be noted that this change should also fix a bug that's probably not detected, namely the construction of a roof over something built under e.g. a destroyed staircase, as the staircase destruction places t_open_air_rooved above the removed staircase when done indoors, and so the roof placement code would not plug the hole as there's the wrong kind of air there. This has not been tested, however.

A potential future usage of the flag is to indicate that placement of terrain with this flag should trigger a check for a containment breach where .e.g. air is released from an underwater facility when the roof is breached and water should flood the facility, or the air of a spacecraft should be vented.
Note, though, that you'd still have to have additional checks for when walls are breached, as those would presumably still have floors.

### CODE REVIEW

File: src/construction.cpp [hunk 1]
```diff
@@ -1143,18 +1142,18 @@ void complete_construction( Character *you )
...
         } else {
             here.ter_set( terp, ter_str_id( built.post_terrain ) );
-            // Make a roof if constructed terrain should have it and it's an open air
+            // Make a roof if constructed terrain should have it and it's an open space
             if( construct::check_up_OK( terp ) ) {
                 const int_id<ter_t> post_terrain = ter_id( built.post_terrain );
                 if( post_terrain->roof ) {
                     const tripoint_bub_ms top = terp + tripoint_above;
-                    if( here.ter( top ) == ter_t_open_air ) {
+                    if( here.ter( top ).id()->has_flag( "EMPTY_SPACE" ) ) {
```
Comment (anothersimulacrum) [inline src/construction.cpp:1150]: You don't need to convert to a string id first.
```suggestion
                    if( here.ter( top )->has_flag( "EMPTY_SPACE" ) ) {
```

File: src/construction.cpp [hunk 2]
```diff
@@ -1143,18 +1142,18 @@ void complete_construction( Character *you )
...
                         here.ter_set( top, ter_id( post_terrain->roof ) );
                     }
                 }
             }
 
-            if( ter_id( built.post_terrain ).id() == ter_t_open_air ) {
+            if( ter_id( built.post_terrain ).id()->has_flag( "EMPTY_SPACE" ) ) {
```
Comment (anothersimulacrum) [inline src/construction.cpp:1156]: Ditto, even though it was existing.
```suggestion
            if( ter_id( built.post_terrain )->has_flag( "EMPTY_SPACE" ) ) {
```

File: src/map.cpp [hunk 3]
```diff
@@ -4250,7 +4250,7 @@ void map::bash_ter_furn( const tripoint &p, bash_params &params )
         spawn_items( p, item_group::items_from( bash->drop_group, calendar::turn ) );
     }
 
-    if( smash_ter && ter( p ) == ter_t_open_air && zlevels ) {
+    if( smash_ter && ter( p ).id()->has_flag( "EMPTY_SPACE" ) && zlevels ) {
```
Comment (anothersimulacrum) [inline src/map.cpp:4253]:
```suggestion
    if( smash_ter && ter( p )->has_flag( "EMPTY_SPACE" ) && zlevels ) {
```

File: src/map.cpp [hunk 4]
```diff
@@ -8719,7 +8719,7 @@ void map::add_roofs( const tripoint &grid )
     for( int x = 0; x < SEEX; x++ ) {
         for( int y = 0; y < SEEY; y++ ) {
             const ter_id ter_here = sub_here->get_ter( { x, y } );
-            if( ter_here != ter_t_open_air ) {
+            if( !ter_here.id()->has_flag( "EMPTY_SPACE" ) ) {
```
Comment (anothersimulacrum) [inline src/map.cpp:8722]:
```suggestion
            if( !ter_here->has_flag( "EMPTY_SPACE" ) ) {
```

**Review by anothersimulacrum — COMMENTED** [review]
https://github.com/CleverRaven/Cataclysm-DDA/blob/874122c314d371c0275a38fe0cafb0feb9bca2f3/src/type_id.h#L262
https://github.com/CleverRaven/Cataclysm-DDA/blob/874122c314d371c0275a38fe0cafb0feb9bca2f3/src/map.h#L853-L855
https://github.com/CleverRaven/Cataclysm-DDA/blob/874122c314d371c0275a38fe0cafb0feb9bca2f3/src/mapdata.cpp#L50-L53

**Comment by PatrikLundell** [comment]
Closing this PR because the same functionality is supposed to be provided by the two flags NO_FLOOR and NO_FLOOR_WATER. It can be noted that the latter flag is missing in a number of instances.
I intend to open a new PR to deal with that as well as the usages covered above.

**Comment by PatrikLundell** [comment]
Reopen the PR, because it turns out NO_FLOOR actually means there is not support beneath to support a person, not that the space is empty (wasp paper floor variant and skylight).
I also won't get into the NO_FLOOR_WATER/DEEP_WATER distinction (if there is any) or whether NO_FLOOR_WATER flags are missing from some deep water or not.

**Pull request merged**